cudax/include/cuda/experimental/detail/hierarchy_dimensions.cuh

```diff
@@ -0,0 +1,811 @@
+  {
+    auto selected = levels_range<Unit, Level>();
+    return detail::convert_to_query_result(::cuda::std::apply(detail::hierarchy_extents_helper<Unit>{}, selected));
+  }
+
+  /**
+   * @brief Returns a count of specified entities at a level in this hierarchy.
+   *
+   * This function return a product of all dimensions of each level in the range between Level and Unit.
+   * Unit and Level need to be levels present in this hierarchy.
+   *
+   * @par Snippet
+   * @code
+   * #include <cudax/hierarchy_dimensions.cuh>
+   * #include <cassert>
+   *
+   * using namespace cuda::experimental;
+   *
+   * auto hierarchy = make_hierarchy(grid_dims(256), cluster_dims<4>(), block_dims<8, 8, 8>());
+   * static_assert(hierarchy.count(thread, cluster) == 4 * 8 * 8 * 8);
+   *
+   * // Using default arguments:
+   * assert(hierarchy.count() == 256 * 4 * 8 * 8 * 8);
+   * assert(hierarchy.count(cluster) == 256);
+   * @endcode
+   * @par
+   *
+   * @tparam Unit
+   *  Specifies what should be counted
+   *
+   * @tparam Level
+   *  Specifies at what level the count should happen
+   */
+  template <typename Unit  = BottomUnit,
+            typename Level = typename detail::get_first_level_type<Levels...>::type::level_type>
+  _CCCL_HOST_DEVICE constexpr auto count(const Unit& = Unit(), const Level& = Level()) const noexcept
+  {
+    return detail::dims_to_count(extents<Unit, Level>());
+  }
+
+  // TODO static extents?
+
+  /**
+   * @brief Returns a compile time count of specified entities at a level in this hierarchy type.
+   *
+   * This function return a product of all dimensions of each level in the range between Level and Unit,
+   * if all of those dimensions are specified statically. If at least one of them is a dynamic value,
+   * this function returns cuda::std::dynamic_extent instead.
+   * Unit and Level need to be levels present in this hierarchy.
+   *
+   * @par Snippet
+   * @code
+   * #include <cudax/hierarchy_dimensions.cuh>
+   * #include <cassert>
+   *
+   * using namespace cuda::experimental;
+   *
+   * auto hierarchy = make_hierarchy(grid_dims(256), cluster_dims<4>(), block_dims<8, 8, 8>());
+   * static_assert(hierarchy.static_count(thread, cluster) == 4 * 8 * 8 * 8);
+   *
+   * // Using default arguments:
+   * assert(hierarchy.static_count() == cuda::std::dynamic_extent);
+   * @endcode
+   * @par
+   *
+   * @tparam Unit
+   *  Specifies what should be counted
+   *
+   * @tparam Level
+   *  Specifies at what level the count should happen
+   */
+  template <typename Unit  = BottomUnit,
+            typename Level = typename detail::get_first_level_type<Levels...>::type::level_type>
+  _CCCL_HOST_DEVICE constexpr static auto static_count(const Unit& = Unit(), const Level& = Level()) noexcept
+  {
+    using extents_type = decltype(::cuda::std::declval<hierarchy_dimensions_fragment<BottomUnit, Levels...>>()
+                                    .template extents<Unit, Level>());
```

Comment:
I believe this should be an alias within `hierarchy_dimensions_fragment` rather than the decltype of a function

Reply:
Great idea, moved it to a member alias